cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -4,123 +4,164 @@
+   * Refresh the metadata. Synchronized to prevent concurrent updates to the metadata cache
    */
-  public synchronized ClusterAndGeneration refreshMetadata(long timeout) {
+  public synchronized ClusterAndGeneration refreshMetadata(int timeout) {
     // Do not update metadata if the metadata has just been refreshed.
-    if (_time.milliseconds() >= _metadata.lastSuccessfulUpdate() + _metadataTTL) {
-      // Cruise Control always fetch metadata for all the topics.
-      _metadata.needMetadataForAllTopics(true);
-      int version = _metadata.requestUpdate();
-      long remaining = timeout;
-      Cluster beforeUpdate = _metadata.fetch();
-      boolean isMetadataUpdated = _metadata.version() > version;
-      while (!isMetadataUpdated && remaining > 0) {
-        _metadata.requestUpdate();
-        long start = _time.milliseconds();
-        _networkClient.poll(remaining, start);
-        remaining -= (_time.milliseconds() - start);
-        isMetadataUpdated = _metadata.version() > version;
+    if (_time.milliseconds() >= _lastSuccessfulUpdate + _metadataTTL) {
+      // Cruise Control always fetches metadata for all the topics.
+      boolean isMetadataUpdated = false;
+      Optional<Cluster> refreshedCluster = Optional.empty();
+      try {
+        refreshedCluster = Optional.of(doRefreshMetadata(timeout));
+        isMetadataUpdated = true;
+      } catch (ExecutionException | TimeoutException | InterruptedException e) {
+        LOG.warn("Exception while updating metadata", e);
       }
-      if (isMetadataUpdated) {
+
+      if (isMetadataUpdated && refreshedCluster.isPresent()) {
+        _lastSuccessfulUpdate = _time.milliseconds();
+        _version += 1;
         if (LOG.isDebugEnabled()) {
-          LOG.debug("Updated metadata {}", _metadata.fetch());
+          LOG.debug("Updated metadata {}", _cluster);
         }
-        if (MonitorUtils.metadataChanged(beforeUpdate, _metadata.fetch())) {
+        Cluster newCluster = refreshedCluster.get();
+        if (MonitorUtils.metadataChanged(_cluster, newCluster)) {
           _metadataGeneration.incrementAndGet();
+          _cluster = newCluster;
         }
       } else {
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+                 timeout, _version, _lastSuccessfulUpdate);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeout) throws InterruptedException, ExecutionException, TimeoutException {
```

Comment:
Looks like retrieving `Cluster` requires 3X more NW round trip with this patch compared to the current way of getting metadata -- is it possible to reduce this overhead?

Reply:
I can't think of a way to reduce the NW round trips unless we are okay with not getting the latest data each time. e.g the DescribeCluster call can be cached for less frequent calls.

Do you think 2 extra round trips are a blocker? I would imagine CC is deployed close to the Kafka cluster and round trips would be pretty fast.

I was wondering why we default to refreshing the metadata every 10 seconds, both within the LoadMonitor and the executor (`EXECUTION_PROGRESS_CHECK_INTERVAL_MS_CONFIG` default).
Is that not too often anyway?

Reply:
We could consider extending Kafka to support a `describeTopics()` which describes all present topics, including internal ones

Reply:
I think it is better to use the public, supported API that guarantees compatibility.

If we think the public API isn't efficient enough and we need a single API call that will return the list of topics, their descriptions and the cluster description, we can open an improvement KIP to add this to Apache Kafka. 

I agree with Stan that this doesn't have significant impact on Cruise Control users or operations. Right now, the time consuming part is the generation of proposals. And as Stan said, we can reduce the impact farther by reducing the time between metadata refreshes.

Reply:
I definitely support/prefer using public APIs when possible. That being said, I have concerns on
1. Consistency of the aggregated metadata,
2. Missing/erroneous `offline replicas` information, and
3. (Now less of a concern) Additional NW cost of retrieving `Cluster`.

I believe ideally, we should potentially consider
* using `PartitionInfo` rather than `TopicPartitionInfo` in `TopicDescription` to capture offline replicas, 
* sticky routing to pick nodes across different calls to admin client inside `doRefreshMetadata`, and 
* considering addition of a one-shot `Cluster` retrieval API in `AdminClient`.